src/functions_math.jl

```diff
@@ -22,17 +22,38 @@ function matrix_prod_names(a, b)
     compile_time_return_hack(res)
 end
 
+
 for (NA, NB) in ((1,2), (2,1), (2,2))  #Vector * Vector, is not allowed
     @eval function Base.:*(a::NamedDimsArray{A,T,$NA}, b::NamedDimsArray{B,S,$NB}) where {A,B,T,S}
         L = matrix_prod_names(A,B)
         data = *(parent(a), parent(b))
         return NamedDimsArray{L}(data)
     end
+end
 
-    @eval function Base.:*(a::NamedDimsArray{L,T,$NA}, b::AbstractArray{S,$NB}) where {L,T,S}
-        return *(a, NamedDimsArray{names(b)}(b))
-    end
-    @eval function Base.:*(a::AbstractArray{T,$NA}, b::NamedDimsArray{L,S,$NB}) where {L,T,S}
-        return *(NamedDimsArray{names(a)}(a), b)
+"""
+    @declare_matmul(MatrixT, VectorT=nothing)
+
+This macro helps define matrix multiplication for the types
+with 2D type parameterization `MatrixT` and 1D `VectorT`.
+It defines the various overloads for `Base.:*` that are required.
+It should be used at the top level of a module.
+"""
+macro declare_matmul(MatrixT, VectorT=nothing)
+    dim_combos = VectorT === nothing ? ((2,2),) : ((1,2), (2,1), (2,2))
+    codes = map(dim_combos) do (NA, NB)
+        TA_named = :(NamedDimsArray{<:Any, <:Any, $NA})
+        TB_named = :(NamedDimsArray{<:Any, <:Any, $NB})
+        TA_other = (VectorT, MatrixT)[NA]
+        TB_other = (VectorT, MatrixT)[NB]
+
+        quote
+            Base.:*(a::$TA_named, b::$TB_other) = *(a, NamedDimsArray{names(b)}(b))
```

Comment:
I doubt that this macro will be used anywhere but in this package, but it would be safer to do `NamedDims.NamedDimsArray` just in case.

Reply:
It might actually.
Since it is required to give compat between two wrapper matrix types.